netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -268,70 +270,125 @@ public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Excep
-         * using -1 here, it forces the SSL implementation to treat it as invalid.
-         */
-        host = authority;
-        port = -1;
-      }
-      return new HostPort(host, port);
+    @Override
+    public ChannelHandler newHandler(GrpcHttp2ConnectionHandler grpcHandler) {
+      ChannelHandler gnh = new GrpcNegotiationHandler(grpcHandler);
+      ChannelHandler cth = new ClientTlsHandler(gnh, sslContext, grpcHandler.getAuthority());
+      WaitUntilActiveHandler wuah = new WaitUntilActiveHandler(cth);
+      return wuah;
     }
 
     @Override
-    public ChannelHandler newHandler(GrpcHttp2ConnectionHandler handler) {
-      final HostPort hostPort = parseAuthority(handler.getAuthority());
+    public void close() {}
+  }
 
-      ChannelHandler sslBootstrap = new ChannelHandlerAdapter() {
-        @Override
-        public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
-          SSLEngine sslEngine = sslContext.newEngine(ctx.alloc(), hostPort.host, hostPort.port);
-          SSLParameters sslParams = sslEngine.getSSLParameters();
-          sslParams.setEndpointIdentificationAlgorithm("HTTPS");
-          sslEngine.setSSLParameters(sslParams);
-          ctx.pipeline().replace(this, null, new SslHandler(sslEngine, false));
-        }
-      };
-      return new BufferUntilTlsNegotiatedHandler(sslBootstrap, handler);
+  static final class ClientTlsHandler extends ChannelDuplexHandler {
+
+    private final ChannelHandler next;
+    private final SslContext sslContext;
+    private final String host;
+    private final int port;
+
+    private ProtocolNegotiationEvent pne = ProtocolNegotiationEvent.DEFAULT;
+
+    ClientTlsHandler(ChannelHandler next, SslContext sslContext, String authority) {
+      this.next = checkNotNull(next, "next");
+      this.sslContext = checkNotNull(sslContext, "sslContext");
+      HostPort hostPort = parseAuthority(authority);
+      this.host = hostPort.host;
+      this.port = hostPort.port;
     }
 
     @Override
-    public AsciiString scheme() {
-      return Utils.HTTPS;
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+      SSLEngine sslEngine = sslContext.newEngine(ctx.alloc(), host, port);
+      SSLParameters sslParams = sslEngine.getSSLParameters();
+      sslParams.setEndpointIdentificationAlgorithm("HTTPS");
+      sslEngine.setSSLParameters(sslParams);
+      ctx.pipeline().addBefore(ctx.name(), /* name= */ null, new SslHandler(sslEngine, false));
+      super.handlerAdded(ctx);
     }
 
     @Override
-    public void close() {}
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+      if (evt instanceof ProtocolNegotiationEvent) {
+        pne = (ProtocolNegotiationEvent) evt;
+      } else if (evt instanceof SslHandshakeCompletionEvent) {
+        SslHandshakeCompletionEvent handshakeEvent = (SslHandshakeCompletionEvent) evt;
```

Comment:
the ProtocolNegotiationEvent must be fired before SslHandshakeCompletionEvent? if it sees the SslHandshakeCompletionEvent before ProtocolNegotiationEvent, should it fail?

Reply:
It may, but I wanted to avoid too much state in the handler.  The idea for these handlers is that they are pluggable and single purpose.  In that regard, they don't assume that a previous PNE, or else they would assume a previous handler, which is tight coupling.

PNE cannot be fired after SHCE, because it removed this handler from the pipeline.  Thus, the invariant is held that if there is is a PNE, it must happen before the SHCE.

Reply:
okay, i see the point. keeping default to null vs default attribute doesn't seems like having much of state than allows unexpected behavior in some weird race condition. since i don't see the race can happen, i am fine with that.